src/java.desktop/share/classes/sun/font/HBShaper.java

```diff
@@ -0,0 +1,630 @@
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.lang.invoke.VarHandle;
+
+import java.util.Optional;
+import java.util.WeakHashMap;
+
+public class HBShaper {
+
+    /*
+     * union _hb_var_int_t {
+     *     uint32_t u32;
+     *     int32_t i32;
+     *     uint16_t u16[2];
+     *     int16_t i16[2];
+     *     uint8_t u8[4];
+     *     int8_t i8[4];
+     * };
+     */
+   private static final UnionLayout VarIntLayout = MemoryLayout.unionLayout(
+        JAVA_INT.withName("u32"),
+        JAVA_INT.withName("i32"),
+        MemoryLayout.sequenceLayout(2, JAVA_SHORT).withName("u16"),
+        MemoryLayout.sequenceLayout(2, JAVA_SHORT).withName("i16"),
+        MemoryLayout.sequenceLayout(4, JAVA_BYTE).withName("u8"),
+        MemoryLayout.sequenceLayout(4, JAVA_BYTE).withName("i8")
+    ).withName("_hb_var_int_t");
+
+    /*
+     * struct hb_glyph_position_t {
+     *     hb_position_t x_advance;
+     *     hb_position_t y_advance;
+     *     hb_position_t x_offset;
+     *     hb_position_t y_offset;
+     *     hb_var_int_t var;
+     * };
+     */
+     private static final StructLayout PositionLayout = MemoryLayout.structLayout(
+        JAVA_INT.withName("x_advance"),
+        JAVA_INT.withName("y_advance"),
+        JAVA_INT.withName("x_offset"),
+        JAVA_INT.withName("y_offset"),
+        VarIntLayout.withName("var")
+     ).withName("hb_glyph_position_t");
+
+    /**
+     * struct hb_glyph_info_t {
+     *     hb_codepoint_t codepoint;
+     *     hb_mask_t mask;
+     *     uint32_t cluster;
+     *     hb_var_int_t var1;
+     *     hb_var_int_t var2;
+     * };
+     */
+    private static final StructLayout GlyphInfoLayout = MemoryLayout.structLayout(
+        JAVA_INT.withName("codepoint"),
+        JAVA_INT.withName("mask"),
+        JAVA_INT.withName("cluster"),
+        VarIntLayout.withName("var1"),
+        VarIntLayout.withName("var2")
+    ).withName("hb_glyph_info_t");
+
+    private static VarHandle getVarHandle(MemoryLayout layout, String name) {
+        return layout.varHandle(
+            PathElement.sequenceElement(),
+            PathElement.groupElement(name));
+    }
+
+    private static VarHandle x_offsetHandle;
+    private static VarHandle y_offsetHandle;
+    private static VarHandle x_advanceHandle;
+    private static VarHandle y_advanceHandle;
+    private static VarHandle codePointHandle;
+    private static VarHandle clusterHandle;
+
+    private static MethodHandles.Lookup MH_LOOKUP;
+    private static Linker LINKER;
+    private static SymbolLookup SYM_LOOKUP;
+    private static MethodHandle malloc_handle;
+    private static MethodHandle create_face_handle;
+    private static MethodHandle dispose_face_handle;
+    private static MethodHandle jdk_hb_shape_handle;
+
+    private static FunctionDescriptor get_nominal_glyph_fd;
+    private static MethodHandle get_nominal_glyph_mh;
```

Comment:
Declaring all these `final` would improve performance.  For example

`private static final FunctionDescription GET_NOMINAL_GLYPH_MH;`